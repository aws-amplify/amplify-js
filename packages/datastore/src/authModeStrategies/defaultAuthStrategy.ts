// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { AuthModeStrategy } from '../types';

// Default behavior is to use the primary auth mode for an API,
// so we are returning an empty array so that DataStore will default
// to using the primary auth mode.
export const defaultAuthStrategy: AuthModeStrategy = () => [];
