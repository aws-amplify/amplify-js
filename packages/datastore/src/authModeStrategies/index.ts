// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export { defaultAuthStrategy } from './defaultAuthStrategy';
export { multiAuthStrategy } from './multiAuthStrategy';
