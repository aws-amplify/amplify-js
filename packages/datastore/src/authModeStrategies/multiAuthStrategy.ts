// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { fetchAuthSession } from '@aws-amplify/core';
import { GraphQLAuthMode } from '@aws-amplify/core/internals/utils';

import {
	AmplifyContext,
	AuthModeStrategy,
	ModelAttributeAuthAllow,
	ModelAttributeAuthProperty,
	ModelAttributeAuthProvider,
} from '../types';

function getProviderFromRule(
	rule: ModelAttributeAuthProperty,
): ModelAttributeAuthProvider {
	// private with no provider means userPools
	if (rule.allow === 'private' && !rule.provider) {
		return ModelAttributeAuthProvider.USER_POOLS;
	}
	// public with no provider means apiKey
	if (rule.allow === 'public' && !rule.provider) {
		return ModelAttributeAuthProvider.API_KEY;
	}

	return rule.provider!;
}

function sortAuthRulesWithPriority(rules: ModelAttributeAuthProperty[]) {
	const allowSortPriority = [
		ModelAttributeAuthAllow.CUSTOM,
		ModelAttributeAuthAllow.OWNER,
		ModelAttributeAuthAllow.GROUPS,
		ModelAttributeAuthAllow.PRIVATE,
		ModelAttributeAuthAllow.PUBLIC,
	];
	const providerSortPriority = [
		ModelAttributeAuthProvider.FUNCTION,
		ModelAttributeAuthProvider.USER_POOLS,
		ModelAttributeAuthProvider.OIDC,
		ModelAttributeAuthProvider.IAM,
		ModelAttributeAuthProvider.API_KEY,
	];

	return [...rules].sort(
		(a: ModelAttributeAuthProperty, b: ModelAttributeAuthProperty) => {
			if (a.allow === b.allow) {
				return (
					providerSortPriority.indexOf(getProviderFromRule(a)) -
					providerSortPriority.indexOf(getProviderFromRule(b))
				);
			}

			return (
				allowSortPriority.indexOf(a.allow) - allowSortPriority.indexOf(b.allow)
			);
		},
	);
}

function getAuthRules({
	rules,
	currentUser,
}: {
	rules: ModelAttributeAuthProperty[];
	currentUser: unknown;
}) {
	// Using Set to ensure uniqueness
	const authModes = new Set<GraphQLAuthMode>();

	rules.forEach(rule => {
		switch (rule.allow) {
			case ModelAttributeAuthAllow.CUSTOM:
				// custom with no provider -> function
				if (
					!rule.provider ||
					rule.provider === ModelAttributeAuthProvider.FUNCTION
				) {
					authModes.add('lambda');
				}
				break;
			case ModelAttributeAuthAllow.GROUPS:
			case ModelAttributeAuthAllow.OWNER: {
				// We shouldn't attempt User Pool or OIDC if there isn't an authenticated user
				if (currentUser) {
					if (rule.provider === ModelAttributeAuthProvider.USER_POOLS) {
						authModes.add('userPool');
					} else if (rule.provider === ModelAttributeAuthProvider.OIDC) {
						authModes.add('oidc');
					}
				}
				break;
			}
			case ModelAttributeAuthAllow.PRIVATE: {
				// We shouldn't attempt private if there isn't an authenticated user
				if (currentUser) {
					// private with no provider means userPools
					if (
						!rule.provider ||
						rule.provider === ModelAttributeAuthProvider.USER_POOLS
					) {
						authModes.add('userPool');
					} else if (rule.provider === ModelAttributeAuthProvider.IAM) {
						authModes.add('iam');
					}
				}

				break;
			}
			case ModelAttributeAuthAllow.PUBLIC: {
				if (rule.provider === ModelAttributeAuthProvider.IAM) {
					authModes.add('iam');
				} else if (
					!rule.provider ||
					rule.provider === ModelAttributeAuthProvider.API_KEY
				) {
					// public with no provider means apiKey
					authModes.add('apiKey');
				}
				break;
			}
			default:
				break;
		}
	});

	return Array.from(authModes);
}

/**
 * Returns an array of auth modes to try based on the schema, model, and
 * authenticated user (or lack thereof). Rules are sourced from `getAuthRules`
 * and returned in the order they ought to be attempted.
 *
 * @see sortAuthRulesWithPriority
 * @see getAuthRules
 *
 * @param param0 The `{schema, modelName}` to inspect.
 * @returns A sorted array of auth modes to attempt.
 */
export const multiAuthStrategy: (
	amplifyContext: AmplifyContext,
) => AuthModeStrategy =
	() =>
	async ({ schema, modelName }) => {
		let currentUser;
		try {
			const authSession = await fetchAuthSession();
			if (authSession.tokens.accessToken) {
				// the user is authenticated
				currentUser = authSession;
			}
		} catch (e) {
			// No current user
		}

		const { attributes } = schema.namespaces.user.models[modelName];

		if (attributes) {
			const authAttribute = attributes.find(attr => attr.type === 'auth');

			if (authAttribute?.properties?.rules) {
				const sortedRules = sortAuthRulesWithPriority(
					authAttribute.properties.rules,
				);

				return getAuthRules({ currentUser, rules: sortedRules });
			}
		}

		return [];
	};
