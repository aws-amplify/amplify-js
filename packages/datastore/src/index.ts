// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import {
	USER,
	isModelConstructor,
	isNonModelConstructor,
	traverseModel,
	validatePredicate,
} from './util';

export {
	DataStore,
	DataStoreClass,
	initSchema,
	ModelInstanceCreator,
	// eslint-disable-next-line import/export
	AsyncCollection,
	AsyncItem,
} from './datastore/datastore';

export {
	Predicates,
	ModelPredicateCreator,
	ModelSortPredicateCreator,
} from './predicates';
export { Adapter as StorageAdapter } from './storage/adapter';

export { NAMESPACES } from './util';

export const utils = {
	USER,
	traverseModel,
	validatePredicate,
	isNonModelConstructor,
	isModelConstructor,
};

// eslint-disable-next-line import/export
export * from './types';
