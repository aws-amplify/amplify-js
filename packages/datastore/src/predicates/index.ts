// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import {
	ModelPredicate,
	PersistentModel,
	PredicatesGroup,
	ProducerModelPredicate,
	SchemaModel,
} from '../types';
import { extractPrimaryKeyFieldNames, extractPrimaryKeyValues } from '../util';

export { ModelSortPredicateCreator } from './sort';

const predicatesAllSet = new WeakSet<ProducerModelPredicate<any>>();

export function isPredicatesAll(
	predicate: any,
): predicate is typeof PredicateAll {
	return predicatesAllSet.has(predicate);
}

/**
 * The valid logical grouping keys for a predicate group.
 */
const groupKeys = new Set(['and', 'or', 'not']);

/**
 * Determines whether an object is a GraphQL style predicate "group", which must be an
 * object containing a single "group key", which then contains the child condition(s).
 *
 * E.g.,
 *
 * ```
 * { and: [ ... ] }
 * { not: { ... } }
 * ```
 *
 * @param o The object to test.
 */
const isGroup = o => {
	const keys = [...Object.keys(o)];

	return keys.length === 1 && groupKeys.has(keys[0]);
};

/**
 * Determines whether an object specifies no conditions and should match everything,
 * as would be the case with `Predicates.ALL`.
 *
 * @param o The object to test.
 */
const isEmpty = o => {
	return !Array.isArray(o) && Object.keys(o).length === 0;
};

/**
 * The valid comparison operators that can be used as keys in a predicate comparison object.
 */
export const comparisonKeys = new Set([
	'eq',
	'ne',
	'gt',
	'lt',
	'ge',
	'le',
	'contains',
	'notContains',
	'beginsWith',
	'between',
]);

/**
 * Determines whether an object is a GraphQL style predicate comparison node, which must
 * be an object containing a single "comparison operator" key, which then contains the
 * operand or operands to compare against.
 *
 * @param o The object to test.
 */
const isComparison = o => {
	const keys = [...Object.keys(o)];

	return !Array.isArray(o) && keys.length === 1 && comparisonKeys.has(keys[0]);
};

/**
 * A light check to determine whether an object is a valid GraphQL Condition AST.
 *
 * @param o The object to test.
 */
const isValid = o => {
	if (Array.isArray(o)) {
		return o.every(v => isValid(v));
	} else {
		return Object.keys(o).length <= 1;
	}
};

// This symbol is not used at runtime, only its type (unique symbol)
export const PredicateAll = Symbol('A predicate that matches all records');

export class Predicates {
	public static get ALL(): typeof PredicateAll {
		// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
		const predicate = <ProducerModelPredicate<any>>(c => c);

		predicatesAllSet.add(predicate);

		return predicate as unknown as typeof PredicateAll;
	}
}

export class ModelPredicateCreator {
	/**
	 * Map of storage predicates (key objects) to storage predicate AST's.
	 */
	private static predicateGroupsMap = new WeakMap<
		ModelPredicate<any>,
		PredicatesGroup<any>
	>();

	/**
	 * Determines whether the given storage predicate (lookup key) is a predicate
	 * key that DataStore recognizes.
	 *
	 * @param predicate The storage predicate (lookup key) to test.
	 */
	static isValidPredicate<T extends PersistentModel>(
		predicate: any,
	): predicate is ModelPredicate<T> {
		return ModelPredicateCreator.predicateGroupsMap.has(predicate);
	}

	/**
	 * Looks for the storage predicate AST that corresponds to a given storage
	 * predicate key.
	 *
	 * The key must have been created internally by a DataStore utility
	 * method, such as `ModelPredicate.createFromAST()`.
	 *
	 * @param predicate The predicate reference to look up.
	 * @param throwOnInvalid Whether to throw an exception if the predicate
	 * isn't a valid DataStore predicate.
	 */
	static getPredicates<T extends PersistentModel>(
		predicate: ModelPredicate<T>,
		throwOnInvalid = true,
	) {
		if (throwOnInvalid && !ModelPredicateCreator.isValidPredicate(predicate)) {
			throw new Error('The predicate is not valid');
		}

		return ModelPredicateCreator.predicateGroupsMap.get(predicate);
	}

	/**
	 * using the PK values from the given `model` (which can be a partial of T
	 * Creates a predicate that matches an instance described by `modelDefinition`
	 * that contains only PK field values.)
	 *
	 * @param modelDefinition The model definition to create a predicate for.
	 * @param model The model instance to extract value equalities from.
	 */
	static createForPk<T extends PersistentModel>(
		modelDefinition: SchemaModel,
		model: T,
	) {
		const keyFields = extractPrimaryKeyFieldNames(modelDefinition);
		const keyValues = extractPrimaryKeyValues(model, keyFields);

		const predicate = this.createFromAST<T>(modelDefinition, {
			and: keyFields.map((field, idx) => {
				const operand = keyValues[idx];

				return { [field]: { eq: operand } };
			}),
		});

		return predicate;
	}

	/**
	 * Searches a `Model` table for records matching the given equalities object.
	 *
	 * This only matches against fields given in the equalities object. No other
	 * fields are tested by the predicate.
	 *
	 * @param modelDefinition The model we need a predicate for.
	 * @param flatEqualities An object holding field equalities to search for.
	 */
	static createFromFlatEqualities<T extends PersistentModel>(
		modelDefinition: SchemaModel,
		flatEqualities: Record<string, any>,
	) {
		const ast = {
			and: Object.entries(flatEqualities).map(([k, v]) => ({ [k]: { eq: v } })),
		};

		return this.createFromAST<T>(modelDefinition, ast);
	}

	/**
	 * Accepts a GraphQL style filter predicate tree and transforms it into an
	 * AST that can be used for a storage adapter predicate. Example input:
	 *
	 * ```js
	 * {
	 * 	and: [
	 * 		{ name: { eq: "Bob Jones" } },
	 * 		{ age: { between: [32, 64] } },
	 * 		{ not: {
	 * 			or: [
	 * 				{ favoriteFood: { eq: 'pizza' } },
	 * 				{ favoriteFood: { eq: 'tacos' } },
	 * 			]
	 * 		}}
	 * 	]
	 * }
	 * ```
	 *
	 * @param gql GraphQL style filter node.
	 */
	static transformGraphQLFilterNodeToPredicateAST(gql: any) {
		if (!isValid(gql)) {
			throw new Error(
				'Invalid GraphQL Condition or subtree: ' + JSON.stringify(gql),
			);
		}

		if (isEmpty(gql)) {
			return {
				type: 'and',
				predicates: [],
			};
		} else if (isGroup(gql)) {
			const groupkey = Object.keys(gql)[0];
			const children = this.transformGraphQLFilterNodeToPredicateAST(
				gql[groupkey],
			);

			return {
				type: groupkey,
				predicates: Array.isArray(children) ? children : [children],
			};
		} else if (isComparison(gql)) {
			const operatorKey = Object.keys(gql)[0];

			return {
				operator: operatorKey,
				operand: gql[operatorKey],
			};
		} else {
			if (Array.isArray(gql)) {
				return gql.map(o => this.transformGraphQLFilterNodeToPredicateAST(o));
			} else {
				const fieldKey = Object.keys(gql)[0];

				return {
					field: fieldKey,
					...this.transformGraphQLFilterNodeToPredicateAST(gql[fieldKey]),
				};
			}
		}
	}

	/**
	 * Accepts a GraphQL style filter predicate tree and transforms it into a predicate
	 * that storage adapters understand. Example input:
	 *
	 * ```js
	 * {
	 * 	and: [
	 * 		{ name: { eq: "Bob Jones" } },
	 * 		{ age: { between: [32, 64] } },
	 * 		{ not: {
	 * 			or: [
	 * 				{ favoriteFood: { eq: 'pizza' } },
	 * 				{ favoriteFood: { eq: 'tacos' } },
	 * 			]
	 * 		}}
	 * 	]
	 * }
	 * ```
	 *
	 * @param modelDefinition The model that the AST/predicate must be compatible with.
	 * @param ast The graphQL style AST that should specify conditions for `modelDefinition`.
	 */
	static createFromAST<T extends PersistentModel>(
		modelDefinition: SchemaModel,
		ast: any,
	): ModelPredicate<T> {
		const key = {} as ModelPredicate<T>;

		ModelPredicateCreator.predicateGroupsMap.set(
			key,
			this.transformGraphQLFilterNodeToPredicateAST(ast),
		);

		return key;
	}
}
