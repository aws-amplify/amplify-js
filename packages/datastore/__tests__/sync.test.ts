// These tests should be replaced once SyncEngine.partialDataFeatureFlagEnabled is removed.
import {
	AmplifyError,
	Category,
	DataStoreAction,
} from '@aws-amplify/core/internals/utils';
import { defaultAuthStrategy } from '../src/authModeStrategies';
let mockGraphQl;

const sessionStorageMock = (() => {
	let store = {};

	return {
		getItem(key) {
			return store[key] || null;
		},
		setItem(key, value) {
			store[key] = value.toString();
		},
		removeItem(key) {
			delete store[key];
		},
		clear() {
			store = {};
		},
	};
})();

Object.defineProperty(window, 'sessionStorage', {
	value: sessionStorageMock,
});
const defaultQuery = `query {
			syncPosts {
				items {
					id
					title
					count
					_version
					_lastChangedAt
					_deleted
				}
				nextToken
				startedAt
			}
		}`;
const defaultVariables = {};
const defaultOpName = 'syncPosts';
const defaultModelDefinition = { name: 'Post' };
const defaultAuthMode = 'userPool';

describe('Sync', () => {
	describe('jitteredRetry', () => {
		beforeEach(() => {
			window.sessionStorage.clear();
			jest.resetModules();
			jest.resetAllMocks();
		});

		it('should return all data', async () => {
			const resolveResponse = {
				data: {
					syncPosts: {
						items: [
							{
								id: '1',
								title: 'Item 1',
							},
							{
								id: '2',
								title: 'Item 2',
							},
						],
					},
				},
			};

			const SyncProcessor = jitteredRetrySyncProcessorSetup({
				resolveResponse,
			});

			const data = await SyncProcessor.jitteredRetry({
				query: defaultQuery,
				variables: defaultVariables,
				opName: defaultOpName,
				modelDefinition: defaultModelDefinition,
			});

			expect(data).toMatchSnapshot();
		});

		it('custom pk: should return all data', async () => {
			const resolveResponse = {
				data: {
					syncPosts: {
						items: [
							{
								postId: '1',
								title: 'Item 1',
							},
							{
								postId: '2',
								title: 'Item 2',
							},
						],
					},
				},
			};

			const SyncProcessor = jitteredRetrySyncProcessorSetup({
				resolveResponse,
			});

			const data = await SyncProcessor.jitteredRetry({
				query: defaultQuery,
				variables: defaultVariables,
				opName: defaultOpName,
				modelDefinition: defaultModelDefinition,
			});

			expect(data).toMatchSnapshot();
		});

		it('should return partial data and send Hub event', async () => {
			const rejectResponse = {
				data: {
					syncPosts: {
						items: [
							{
								id: '1',
								title: 'Item 1',
							},
							null,
							{
								id: '3',
								title: 'Item 3',
							},
						],
					},
				},
				errors: [
					{
						message: 'Item 2 error',
					},
				],
			};

			const hubDispatchMock = jest.fn();
			const coreMocks = {
				Hub: {
					dispatch: hubDispatchMock,
					listen: jest.fn(),
				},
			};

			const SyncProcessor = jitteredRetrySyncProcessorSetup({
				rejectResponse,
				coreMocks,
			});

			const data = await SyncProcessor.jitteredRetry({
				query: defaultQuery,
				variables: defaultVariables,
				opName: defaultOpName,
				modelDefinition: defaultModelDefinition,
				authMode: defaultAuthMode,
			});

			expect(data).toMatchSnapshot();

			expect(hubDispatchMock).toHaveBeenCalledWith('datastore', {
				event: 'nonApplicableDataReceived',
				data: {
					errors: [
						{
							message: 'Item 2 error',
						},
					],
					modelName: 'Post',
				},
			});
		});

		// TODO(v6) Re-enable test
		it('should throw error if no data is returned', async () => {
			const rejectResponse = {
				data: null,
				errors: [
					{
						message: 'General error',
					},
				],
			};

			const SyncProcessor = jitteredRetrySyncProcessorSetup({
				rejectResponse,
			});

			try {
				await SyncProcessor.jitteredRetry({
					query: defaultQuery,
					variables: defaultVariables,
					opName: defaultOpName,
					modelDefinition: defaultModelDefinition,
					authMode: defaultAuthMode,
				});
			} catch (e) {
				expect(e).toMatchSnapshot();
			}
		});

		it('should return NonRetryableError for 403 error', async () => {
			const rejectResponse = {
				data: null,
				errors: [
					{
						originalError: {
							$metadata: {
								httpStatusCode: 403,
							},
						},
					},
				],
			};

			const SyncProcessor = jitteredRetrySyncProcessorSetup({
				rejectResponse,
			});

			try {
				await SyncProcessor.jitteredRetry({
					query: defaultQuery,
					variables: defaultVariables,
					opName: defaultOpName,
					modelDefinition: defaultModelDefinition,
					authMode: defaultAuthMode,
				});
			} catch (e) {
				// NonRetryableError has a `nonRetryable` property
				expect(e).toHaveProperty('nonRetryable');
			}
		});

		[
			'No api-key configured',
			'No current user',
			'No credentials',
			'No federated jwt',
		].forEach(authError => {
			it(`should return NonRetryableError for client-side error: ${authError}`, async () => {
				const rejectResponse = {
					message: authError,
				};

				const SyncProcessor = jitteredRetrySyncProcessorSetup({
					rejectResponse,
				});

				try {
					await SyncProcessor.jitteredRetry({
						query: defaultQuery,
						variables: defaultVariables,
						opName: defaultOpName,
						modelDefinition: defaultModelDefinition,
						authMode: defaultAuthMode,
					});
				} catch (e) {
					// NonRetryableError has a `nonRetryable` property
					expect(e).toHaveProperty('nonRetryable');
				}
			});
		});

		it('should send datastore user agent details with graphql request', async () => {
			const resolveResponse = {
				data: {
					syncPosts: {
						items: [
							{
								id: '1',
								title: 'Item 1',
							},
							{
								id: '2',
								title: 'Item 2',
							},
						],
					},
				},
			};

			const SyncProcessor = jitteredRetrySyncProcessorSetup({
				resolveResponse,
			});

			await SyncProcessor.jitteredRetry({
				query: defaultQuery,
				variables: defaultVariables,
				opName: defaultOpName,
				modelDefinition: defaultModelDefinition,
			});

			expect(mockGraphQl).toHaveBeenCalledWith(expect.anything(), undefined, {
				category: Category.DataStore,
				action: DataStoreAction.GraphQl,
			});
		});
	});

	describe('error handler', () => {
		const errorHandler = jest.fn();
		const data = {
			syncPosts: {
				items: [
					{
						id: '1',
						title: 'Item 1',
					},
					null,
					{
						id: '3',
						title: 'Item 3',
					},
				],
			},
		};

		beforeEach(async () => {
			window.sessionStorage.clear();
			jest.resetModules();
			jest.resetAllMocks();
			errorHandler.mockClear();
		});

		test('bad record', async () => {
			const syncProcessor = jitteredRetrySyncProcessorSetup({
				errorHandler,
				rejectResponse: {
					data,
					errors: [
						{
							message: 'Cannot return boolean for string type',
						},
					],
				},
			});

			await syncProcessor.jitteredRetry({
				query: defaultQuery,
				variables: defaultVariables,
				opName: defaultOpName,
				modelDefinition: defaultModelDefinition,
			});

			expect(errorHandler).toHaveBeenCalledWith(
				expect.objectContaining({
					operation: 'syncPosts',
					process: 'sync',
					errorType: 'BadRecord',
				}),
			);
		});

		test('connection timeout', async () => {
			const syncProcessor = jitteredRetrySyncProcessorSetup({
				errorHandler,
				rejectResponse: {
					data,
					errors: [
						{
							message: 'Connection failed: Connection Timeout',
						},
					],
				},
			});

			await syncProcessor.jitteredRetry({
				query: defaultQuery,
				variables: defaultVariables,
				opName: defaultOpName,
				modelDefinition: defaultModelDefinition,
			});

			expect(errorHandler).toHaveBeenCalledWith(
				expect.objectContaining({
					operation: 'syncPosts',
					process: 'sync',
					errorType: 'Transient',
				}),
			);
		});

		test('server error', async () => {
			const syncProcessor = jitteredRetrySyncProcessorSetup({
				errorHandler,
				rejectResponse: {
					data,
					errors: [
						{
							originalError: {
								$metadata: {
									httpStatusCode: 500,
								},
							},
						},
					],
				},
			});

			await syncProcessor.jitteredRetry({
				query: defaultQuery,
				variables: defaultVariables,
				opName: defaultOpName,
				modelDefinition: defaultModelDefinition,
			});

			expect(errorHandler).toHaveBeenCalledWith(
				expect.objectContaining({
					operation: 'syncPosts',
					process: 'sync',
					errorType: 'Transient',
				}),
			);
		});
	});
});

function jitteredRetrySyncProcessorSetup({
	rejectResponse,
	resolveResponse,
	coreMocks,
	errorHandler = () => null,
}: {
	rejectResponse?: any;
	resolveResponse?: any;
	coreMocks?: object;
	errorHandler?: () => null;
}) {
	mockGraphQl = jest.fn(
		() =>
			new Promise((res, rej) => {
				if (resolveResponse) {
					res(resolveResponse);
				} else if (rejectResponse) {
					rej(rejectResponse);
				}
			}),
	);
	// mock graphql to return a mockable observable
	jest.mock('@aws-amplify/api/internals', () => {
		const actualInternalAPIModule = jest.requireActual(
			'@aws-amplify/api/internals',
		);
		const actualInternalAPIInstance = actualInternalAPIModule.InternalAPI;

		return {
			...actualInternalAPIModule,
			InternalAPI: {
				...actualInternalAPIInstance,
				graphql: mockGraphQl,
			},
		};
	});

	jest.mock('@aws-amplify/core/internals/utils', () => ({
		...jest.requireActual('@aws-amplify/core/internals/utils'),
		// No need to retry any thrown errors right now,
		// so we're overriding jitteredExponentialRetry
		jitteredExponentialRetry: (fn, args) => fn(...args),
	}));

	jest.mock('@aws-amplify/core', () => ({
		...jest.requireActual('@aws-amplify/core'),
		...coreMocks,
	}));

	const SyncProcessorClass =
		require('../src/sync/processors/sync').SyncProcessor;

	const testInternalSchema = {
		namespaces: {},
		version: '',
	};

	const SyncProcessor = new SyncProcessorClass(
		testInternalSchema,
		null, // syncPredicates
		{ aws_appsync_authenticationType: 'userPools' },
		defaultAuthStrategy,
		errorHandler,
		{},
	);

	return SyncProcessor;
}
