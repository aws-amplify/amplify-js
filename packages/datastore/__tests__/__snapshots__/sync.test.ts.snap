// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync jitteredRetry custom pk: should return all data 1`] = `
{
  "data": {
    "syncPosts": {
      "items": [
        {
          "postId": "1",
          "title": "Item 1",
        },
        {
          "postId": "2",
          "title": "Item 2",
        },
      ],
    },
  },
}
`;

exports[`Sync jitteredRetry should return all data 1`] = `
{
  "data": {
    "syncPosts": {
      "items": [
        {
          "id": "1",
          "title": "Item 1",
        },
        {
          "id": "2",
          "title": "Item 2",
        },
      ],
    },
  },
}
`;

exports[`Sync jitteredRetry should return partial data and send Hub event 1`] = `
{
  "data": {
    "syncPosts": {
      "items": [
        {
          "id": "1",
          "title": "Item 1",
        },
        {
          "id": "3",
          "title": "Item 3",
        },
      ],
    },
  },
  "errors": [
    {
      "message": "Item 2 error",
    },
  ],
}
`;

exports[`Sync jitteredRetry should throw error if no data is returned 1`] = `
{
  "data": null,
  "errors": [
    {
      "message": "General error",
    },
  ],
}
`;
